File: web/error_codes.py
"""Error codes for C++ Audio Engine integration.

This module provides Python equivalents of C++ error codes defined in
docs/architecture/error-codes.md. All codes are synchronized with
include/error_codes.h for consistent error handling across layers.
"""

from dataclasses import dataclass
from enum import Enum


class ErrorCategory(str, Enum):
    """Error category classification.

    Categories match C++ implementation and are used for:
    - Grouping related errors
    - Determining appropriate HTTP status codes
    - Logging and monitoring purposes
    """

    AUDIO_PROCESSING = "audio_processing"
    DAC_ALSA = "dac_alsa"
    IPC_ZEROMQ = "ipc_zeromq"
    GPU_CUDA = "gpu_cuda"
    VALIDATION = "validation"
    CROSSFEED = "crossfeed"  # #150
    INTERNAL = "internal"


class ErrorCode(str, Enum):
    """Error codes synchronized with C++ AudioEngine::ErrorCode.

    Each code maps to a 16-bit hex value in C++ (e.g., AUDIO_INVALID_INPUT_RATE = 0x1001).
    Python uses string names for readability; mapping to HTTP status is done via ERROR_MAPPINGS.
    """

    # Audio Processing (0x1000)
    AUDIO_INVALID_INPUT_RATE = "AUDIO_INVALID_INPUT_RATE"
    AUDIO_INVALID_OUTPUT_RATE = "AUDIO_INVALID_OUTPUT_RATE"
    AUDIO_UNSUPPORTED_FORMAT = "AUDIO_UNSUPPORTED_FORMAT"
    AUDIO_FILTER_NOT_FOUND = "AUDIO_FILTER_NOT_FOUND"
    AUDIO_BUFFER_OVERFLOW = "AUDIO_BUFFER_OVERFLOW"
    AUDIO_XRUN_DETECTED = "AUDIO_XRUN_DETECTED"
    AUDIO_RTP_SOCKET_ERROR = "AUDIO_RTP_SOCKET_ERROR"
    AUDIO_RTP_SESSION_NOT_FOUND = "AUDIO_RTP_SESSION_NOT_FOUND"

    # DAC/ALSA (0x2000)
    DAC_DEVICE_NOT_FOUND = "DAC_DEVICE_NOT_FOUND"
    DAC_OPEN_FAILED = "DAC_OPEN_FAILED"
    DAC_CAPABILITY_SCAN_FAILED = "DAC_CAPABILITY_SCAN_FAILED"
    DAC_RATE_NOT_SUPPORTED = "DAC_RATE_NOT_SUPPORTED"
    DAC_FORMAT_NOT_SUPPORTED = "DAC_FORMAT_NOT_SUPPORTED"
    DAC_BUSY = "DAC_BUSY"

    # IPC/ZeroMQ (0x3000)
    IPC_CONNECTION_FAILED = "IPC_CONNECTION_FAILED"
    IPC_TIMEOUT = "IPC_TIMEOUT"
    IPC_INVALID_COMMAND = "IPC_INVALID_COMMAND"
    IPC_INVALID_PARAMS = "IPC_INVALID_PARAMS"
    IPC_DAEMON_NOT_RUNNING = "IPC_DAEMON_NOT_RUNNING"
    IPC_PROTOCOL_ERROR = "IPC_PROTOCOL_ERROR"

    # GPU/CUDA (0x4000)
    GPU_INIT_FAILED = "GPU_INIT_FAILED"
    GPU_DEVICE_NOT_FOUND = "GPU_DEVICE_NOT_FOUND"
    GPU_MEMORY_ERROR = "GPU_MEMORY_ERROR"
    GPU_KERNEL_LAUNCH_FAILED = "GPU_KERNEL_LAUNCH_FAILED"
    GPU_FILTER_LOAD_FAILED = "GPU_FILTER_LOAD_FAILED"
    GPU_CUFFT_ERROR = "GPU_CUFFT_ERROR"

    # Validation (0x5000)
    VALIDATION_INVALID_CONFIG = "VALIDATION_INVALID_CONFIG"
    VALIDATION_INVALID_PROFILE = "VALIDATION_INVALID_PROFILE"
    VALIDATION_PATH_TRAVERSAL = "VALIDATION_PATH_TRAVERSAL"
    VALIDATION_FILE_NOT_FOUND = "VALIDATION_FILE_NOT_FOUND"
    VALIDATION_PROFILE_EXISTS = "VALIDATION_PROFILE_EXISTS"
    VALIDATION_INVALID_HEADPHONE = "VALIDATION_INVALID_HEADPHONE"
    ERR_UNSUPPORTED_MODE = "ERR_UNSUPPORTED_MODE"

    # Crossfeed/HRTF (0x6000) - #150
    CROSSFEED_NOT_INITIALIZED = "CROSSFEED_NOT_INITIALIZED"
    CROSSFEED_INVALID_RATE_FAMILY = "CROSSFEED_INVALID_RATE_FAMILY"
    CROSSFEED_NOT_IMPLEMENTED = "CROSSFEED_NOT_IMPLEMENTED"
    CROSSFEED_INVALID_FILTER_SIZE = "CROSSFEED_INVALID_FILTER_SIZE"


@dataclass(frozen=True)
class ErrorMapping:
    """Mapping from error code to HTTP response details."""

    http_status: int
    category: ErrorCategory
    title: str


# Complete mapping of all 30 error codes to HTTP status codes
# Based on docs/architecture/error-codes.md §5
ERROR_MAPPINGS: dict[ErrorCode, ErrorMapping] = {
    # Audio Processing (8 codes)
    ErrorCode.AUDIO_INVALID_INPUT_RATE: ErrorMapping(
        400, ErrorCategory.AUDIO_PROCESSING, "Invalid Input Sample Rate"
    ),
    ErrorCode.AUDIO_INVALID_OUTPUT_RATE: ErrorMapping(
        400, ErrorCategory.AUDIO_PROCESSING, "Invalid Output Sample Rate"
    ),
    ErrorCode.AUDIO_UNSUPPORTED_FORMAT: ErrorMapping(
        400, ErrorCategory.AUDIO_PROCESSING, "Unsupported Audio Format"
    ),
    ErrorCode.AUDIO_FILTER_NOT_FOUND: ErrorMapping(
        404, ErrorCategory.AUDIO_PROCESSING, "Filter Not Found"
    ),
    ErrorCode.AUDIO_BUFFER_OVERFLOW: ErrorMapping(
        500, ErrorCategory.AUDIO_PROCESSING, "Buffer Overflow"
    ),
    ErrorCode.AUDIO_XRUN_DETECTED: ErrorMapping(
        500, ErrorCategory.AUDIO_PROCESSING, "Audio XRUN Detected"
    ),
    ErrorCode.AUDIO_RTP_SOCKET_ERROR: ErrorMapping(
        500, ErrorCategory.AUDIO_PROCESSING, "RTP Socket Error"
    ),
    ErrorCode.AUDIO_RTP_SESSION_NOT_FOUND: ErrorMapping(
        404, ErrorCategory.AUDIO_PROCESSING, "RTP Session Not Found"
    ),
    # DAC/ALSA (6 codes)
    ErrorCode.DAC_DEVICE_NOT_FOUND: ErrorMapping(
        404, ErrorCategory.DAC_ALSA, "DAC Device Not Found"
    ),
    ErrorCode.DAC_OPEN_FAILED: ErrorMapping(
        500, ErrorCategory.DAC_ALSA, "DAC Open Failed"
    ),
    ErrorCode.DAC_CAPABILITY_SCAN_FAILED: ErrorMapping(
        500, ErrorCategory.DAC_ALSA, "DAC Capability Scan Failed"
    ),
    ErrorCode.DAC_RATE_NOT_SUPPORTED: ErrorMapping(
        422, ErrorCategory.DAC_ALSA, "DAC Rate Not Supported"
    ),
    ErrorCode.DAC_FORMAT_NOT_SUPPORTED: ErrorMapping(
        422, ErrorCategory.DAC_ALSA, "DAC Format Not Supported"
    ),
    ErrorCode.DAC_BUSY: ErrorMapping(409, ErrorCategory.DAC_ALSA, "DAC Device Busy"),
    # IPC/ZeroMQ (6 codes)
    ErrorCode.IPC_CONNECTION_FAILED: ErrorMapping(
        503, ErrorCategory.IPC_ZEROMQ, "Daemon Connection Failed"
    ),
    ErrorCode.IPC_TIMEOUT: ErrorMapping(
        504, ErrorCategory.IPC_ZEROMQ, "Daemon Timeout"
    ),
    ErrorCode.IPC_INVALID_COMMAND: ErrorMapping(
        400, ErrorCategory.IPC_ZEROMQ, "Invalid Command"
    ),
    ErrorCode.IPC_INVALID_PARAMS: ErrorMapping(
        400, ErrorCategory.IPC_ZEROMQ, "Invalid Parameters"
    ),
    ErrorCode.IPC_DAEMON_NOT_RUNNING: ErrorMapping(
        503, ErrorCategory.IPC_ZEROMQ, "Daemon Not Running"
    ),
    ErrorCode.IPC_PROTOCOL_ERROR: ErrorMapping(
        500, ErrorCategory.IPC_ZEROMQ, "Protocol Error"
    ),
    # GPU/CUDA (6 codes)
    ErrorCode.GPU_INIT_FAILED: ErrorMapping(
        500, ErrorCategory.GPU_CUDA, "GPU Initialization Failed"
    ),
    ErrorCode.GPU_DEVICE_NOT_FOUND: ErrorMapping(
        500, ErrorCategory.GPU_CUDA, "GPU Device Not Found"
    ),
    ErrorCode.GPU_MEMORY_ERROR: ErrorMapping(
        500, ErrorCategory.GPU_CUDA, "GPU Memory Error"
    ),
    ErrorCode.GPU_KERNEL_LAUNCH_FAILED: ErrorMapping(
        500, ErrorCategory.GPU_CUDA, "GPU Kernel Launch Failed"
    ),
    ErrorCode.GPU_FILTER_LOAD_FAILED: ErrorMapping(
        500, ErrorCategory.GPU_CUDA, "GPU Filter Load Failed"
    ),
    ErrorCode.GPU_CUFFT_ERROR: ErrorMapping(500, ErrorCategory.GPU_CUDA, "cuFFT Error"),
    # Validation (6 codes)
    ErrorCode.VALIDATION_INVALID_CONFIG: ErrorMapping(
        400, ErrorCategory.VALIDATION, "Invalid Configuration"
    ),
    ErrorCode.VALIDATION_INVALID_PROFILE: ErrorMapping(
        400, ErrorCategory.VALIDATION, "Invalid EQ Profile"
    ),
    ErrorCode.VALIDATION_PATH_TRAVERSAL: ErrorMapping(
        400, ErrorCategory.VALIDATION, "Path Traversal Detected"
    ),
    ErrorCode.VALIDATION_FILE_NOT_FOUND: ErrorMapping(
        404, ErrorCategory.VALIDATION, "File Not Found"
    ),
    ErrorCode.VALIDATION_PROFILE_EXISTS: ErrorMapping(
        409, ErrorCategory.VALIDATION, "Profile Already Exists"
    ),
    ErrorCode.VALIDATION_INVALID_HEADPHONE: ErrorMapping(
        404, ErrorCategory.VALIDATION, "Headphone Not Found in OPRA DB"
    ),
    ErrorCode.ERR_UNSUPPORTED_MODE: ErrorMapping(
        400, ErrorCategory.VALIDATION, "Unsupported Output Mode"
    ),
    # Crossfeed/HRTF (4 codes) - #150
    ErrorCode.CROSSFEED_NOT_INITIALIZED: ErrorMapping(
        503, ErrorCategory.CROSSFEED, "Crossfeed Not Initialized"
    ),
    ErrorCode.CROSSFEED_INVALID_RATE_FAMILY: ErrorMapping(
        400, ErrorCategory.CROSSFEED, "Invalid Rate Family"
    ),
    ErrorCode.CROSSFEED_NOT_IMPLEMENTED: ErrorMapping(
        501, ErrorCategory.CROSSFEED, "Crossfeed Filter Application Not Implemented"
    ),
    ErrorCode.CROSSFEED_INVALID_FILTER_SIZE: ErrorMapping(
        400, ErrorCategory.CROSSFEED, "Invalid Filter Size"
    ),
}

# Default mapping for unknown error codes
_DEFAULT_MAPPING = ErrorMapping(500, ErrorCategory.INTERNAL, "Internal Error")


def get_error_mapping(error_code: str) -> ErrorMapping:
    """Get error mapping for a given error code string.

    Args:
        error_code: Error code string (e.g., "DAC_RATE_NOT_SUPPORTED")

    Returns:
        ErrorMapping with http_status, category, and title.
        Returns default 500/INTERNAL mapping for unknown codes.
    """
    try:
        code = ErrorCode(error_code)
        return ERROR_MAPPINGS.get(code, _DEFAULT_MAPPING)
    except ValueError:
        # Unknown error code - return internal error
        return _DEFAULT_MAPPING


def get_category_from_code(error_code: str) -> ErrorCategory:
    """Determine category from error code prefix.

    This is useful when the error code is known but not in ERROR_MAPPINGS.
    Falls back to INTERNAL for unknown prefixes.
    """
    if error_code.startswith("AUDIO_"):
        return ErrorCategory.AUDIO_PROCESSING
    elif error_code.startswith("DAC_"):
        return ErrorCategory.DAC_ALSA
    elif error_code.startswith("IPC_"):
        return ErrorCategory.IPC_ZEROMQ
    elif error_code.startswith("GPU_"):
        return ErrorCategory.GPU_CUDA
    elif error_code.startswith("VALIDATION_"):
        return ErrorCategory.VALIDATION
    elif error_code.startswith("CROSSFEED_"):
        return ErrorCategory.CROSSFEED
    else:
        return ErrorCategory.INTERNAL

File: src/core/error_codes.cpp
#include "core/error_codes.h"

#include <iomanip>
#include <sstream>
#include <unordered_map>

namespace AudioEngine {

// Error code to string mapping
static const std::unordered_map<ErrorCode, const char*> kErrorCodeStrings = {
    {ErrorCode::OK, "OK"},

    // Audio Processing
    {ErrorCode::AUDIO_INVALID_INPUT_RATE, "AUDIO_INVALID_INPUT_RATE"},
    {ErrorCode::AUDIO_INVALID_OUTPUT_RATE, "AUDIO_INVALID_OUTPUT_RATE"},
    {ErrorCode::AUDIO_UNSUPPORTED_FORMAT, "AUDIO_UNSUPPORTED_FORMAT"},
    {ErrorCode::AUDIO_FILTER_NOT_FOUND, "AUDIO_FILTER_NOT_FOUND"},
    {ErrorCode::AUDIO_BUFFER_OVERFLOW, "AUDIO_BUFFER_OVERFLOW"},
    {ErrorCode::AUDIO_XRUN_DETECTED, "AUDIO_XRUN_DETECTED"},

    // DAC/ALSA
    {ErrorCode::DAC_DEVICE_NOT_FOUND, "DAC_DEVICE_NOT_FOUND"},
    {ErrorCode::DAC_OPEN_FAILED, "DAC_OPEN_FAILED"},
    {ErrorCode::DAC_CAPABILITY_SCAN_FAILED, "DAC_CAPABILITY_SCAN_FAILED"},
    {ErrorCode::DAC_RATE_NOT_SUPPORTED, "DAC_RATE_NOT_SUPPORTED"},
    {ErrorCode::DAC_FORMAT_NOT_SUPPORTED, "DAC_FORMAT_NOT_SUPPORTED"},
    {ErrorCode::DAC_BUSY, "DAC_BUSY"},

    // IPC/ZeroMQ
    {ErrorCode::IPC_CONNECTION_FAILED, "IPC_CONNECTION_FAILED"},
    {ErrorCode::IPC_TIMEOUT, "IPC_TIMEOUT"},
    {ErrorCode::IPC_INVALID_COMMAND, "IPC_INVALID_COMMAND"},
    {ErrorCode::IPC_INVALID_PARAMS, "IPC_INVALID_PARAMS"},
    {ErrorCode::IPC_DAEMON_NOT_RUNNING, "IPC_DAEMON_NOT_RUNNING"},
    {ErrorCode::IPC_PROTOCOL_ERROR, "IPC_PROTOCOL_ERROR"},

    // GPU/CUDA
    {ErrorCode::GPU_INIT_FAILED, "GPU_INIT_FAILED"},
    {ErrorCode::GPU_DEVICE_NOT_FOUND, "GPU_DEVICE_NOT_FOUND"},
    {ErrorCode::GPU_MEMORY_ERROR, "GPU_MEMORY_ERROR"},
    {ErrorCode::GPU_KERNEL_LAUNCH_FAILED, "GPU_KERNEL_LAUNCH_FAILED"},
    {ErrorCode::GPU_FILTER_LOAD_FAILED, "GPU_FILTER_LOAD_FAILED"},
    {ErrorCode::GPU_CUFFT_ERROR, "GPU_CUFFT_ERROR"},

    // Validation
    {ErrorCode::VALIDATION_INVALID_CONFIG, "VALIDATION_INVALID_CONFIG"},
    {ErrorCode::VALIDATION_INVALID_PROFILE, "VALIDATION_INVALID_PROFILE"},
    {ErrorCode::VALIDATION_PATH_TRAVERSAL, "VALIDATION_PATH_TRAVERSAL"},
    {ErrorCode::VALIDATION_FILE_NOT_FOUND, "VALIDATION_FILE_NOT_FOUND"},
    {ErrorCode::VALIDATION_PROFILE_EXISTS, "VALIDATION_PROFILE_EXISTS"},
    {ErrorCode::VALIDATION_INVALID_HEADPHONE, "VALIDATION_INVALID_HEADPHONE"},
    {ErrorCode::ERR_UNSUPPORTED_MODE, "ERR_UNSUPPORTED_MODE"},

    // Internal
    {ErrorCode::INTERNAL_UNKNOWN, "INTERNAL_UNKNOWN"},
};

// Error code to HTTP status mapping
static const std::unordered_map<ErrorCode, int> kHttpStatusMap = {
    {ErrorCode::OK, 200},

    // Audio Processing
    {ErrorCode::AUDIO_INVALID_INPUT_RATE, 400},
    {ErrorCode::AUDIO_INVALID_OUTPUT_RATE, 400},
    {ErrorCode::AUDIO_UNSUPPORTED_FORMAT, 400},
    {ErrorCode::AUDIO_FILTER_NOT_FOUND, 404},
    {ErrorCode::AUDIO_BUFFER_OVERFLOW, 500},
    {ErrorCode::AUDIO_XRUN_DETECTED, 500},

    // DAC/ALSA
    {ErrorCode::DAC_DEVICE_NOT_FOUND, 404},
    {ErrorCode::DAC_OPEN_FAILED, 500},
    {ErrorCode::DAC_CAPABILITY_SCAN_FAILED, 500},
    {ErrorCode::DAC_RATE_NOT_SUPPORTED, 422},
    {ErrorCode::DAC_FORMAT_NOT_SUPPORTED, 422},
    {ErrorCode::DAC_BUSY, 409},

    // IPC/ZeroMQ
    {ErrorCode::IPC_CONNECTION_FAILED, 503},
    {ErrorCode::IPC_TIMEOUT, 504},
    {ErrorCode::IPC_INVALID_COMMAND, 400},
    {ErrorCode::IPC_INVALID_PARAMS, 400},
    {ErrorCode::IPC_DAEMON_NOT_RUNNING, 503},
    {ErrorCode::IPC_PROTOCOL_ERROR, 500},

    // GPU/CUDA
    {ErrorCode::GPU_INIT_FAILED, 500},
    {ErrorCode::GPU_DEVICE_NOT_FOUND, 500},
    {ErrorCode::GPU_MEMORY_ERROR, 500},
    {ErrorCode::GPU_KERNEL_LAUNCH_FAILED, 500},
    {ErrorCode::GPU_FILTER_LOAD_FAILED, 500},
    {ErrorCode::GPU_CUFFT_ERROR, 500},

    // Validation
    {ErrorCode::VALIDATION_INVALID_CONFIG, 400},
    {ErrorCode::VALIDATION_INVALID_PROFILE, 400},
    {ErrorCode::VALIDATION_PATH_TRAVERSAL, 400},
    {ErrorCode::VALIDATION_FILE_NOT_FOUND, 404},
    {ErrorCode::VALIDATION_PROFILE_EXISTS, 409},
    {ErrorCode::VALIDATION_INVALID_HEADPHONE, 404},
    {ErrorCode::ERR_UNSUPPORTED_MODE, 400},

    // Internal
    {ErrorCode::INTERNAL_UNKNOWN, 500},
};

// String to error code mapping (reverse lookup)
static const std::unordered_map<std::string, ErrorCode> kStringToErrorCode = {
    {"OK", ErrorCode::OK},

    // Audio Processing
    {"AUDIO_INVALID_INPUT_RATE", ErrorCode::AUDIO_INVALID_INPUT_RATE},
    {"AUDIO_INVALID_OUTPUT_RATE", ErrorCode::AUDIO_INVALID_OUTPUT_RATE},
    {"AUDIO_UNSUPPORTED_FORMAT", ErrorCode::AUDIO_UNSUPPORTED_FORMAT},
    {"AUDIO_FILTER_NOT_FOUND", ErrorCode::AUDIO_FILTER_NOT_FOUND},
    {"AUDIO_BUFFER_OVERFLOW", ErrorCode::AUDIO_BUFFER_OVERFLOW},
    {"AUDIO_XRUN_DETECTED", ErrorCode::AUDIO_XRUN_DETECTED},

    // DAC/ALSA
    {"DAC_DEVICE_NOT_FOUND", ErrorCode::DAC_DEVICE_NOT_FOUND},
    {"DAC_OPEN_FAILED", ErrorCode::DAC_OPEN_FAILED},
    {"DAC_CAPABILITY_SCAN_FAILED", ErrorCode::DAC_CAPABILITY_SCAN_FAILED},
    {"DAC_RATE_NOT_SUPPORTED", ErrorCode::DAC_RATE_NOT_SUPPORTED},
    {"DAC_FORMAT_NOT_SUPPORTED", ErrorCode::DAC_FORMAT_NOT_SUPPORTED},
    {"DAC_BUSY", ErrorCode::DAC_BUSY},

    // IPC/ZeroMQ
    {"IPC_CONNECTION_FAILED", ErrorCode::IPC_CONNECTION_FAILED},
    {"IPC_TIMEOUT", ErrorCode::IPC_TIMEOUT},
    {"IPC_INVALID_COMMAND", ErrorCode::IPC_INVALID_COMMAND},
    {"IPC_INVALID_PARAMS", ErrorCode::IPC_INVALID_PARAMS},
    {"IPC_DAEMON_NOT_RUNNING", ErrorCode::IPC_DAEMON_NOT_RUNNING},
    {"IPC_PROTOCOL_ERROR", ErrorCode::IPC_PROTOCOL_ERROR},

    // GPU/CUDA
    {"GPU_INIT_FAILED", ErrorCode::GPU_INIT_FAILED},
    {"GPU_DEVICE_NOT_FOUND", ErrorCode::GPU_DEVICE_NOT_FOUND},
    {"GPU_MEMORY_ERROR", ErrorCode::GPU_MEMORY_ERROR},
    {"GPU_KERNEL_LAUNCH_FAILED", ErrorCode::GPU_KERNEL_LAUNCH_FAILED},
    {"GPU_FILTER_LOAD_FAILED", ErrorCode::GPU_FILTER_LOAD_FAILED},
    {"GPU_CUFFT_ERROR", ErrorCode::GPU_CUFFT_ERROR},

    // Validation
    {"VALIDATION_INVALID_CONFIG", ErrorCode::VALIDATION_INVALID_CONFIG},
    {"VALIDATION_INVALID_PROFILE", ErrorCode::VALIDATION_INVALID_PROFILE},
    {"VALIDATION_PATH_TRAVERSAL", ErrorCode::VALIDATION_PATH_TRAVERSAL},
    {"VALIDATION_FILE_NOT_FOUND", ErrorCode::VALIDATION_FILE_NOT_FOUND},
    {"VALIDATION_PROFILE_EXISTS", ErrorCode::VALIDATION_PROFILE_EXISTS},
    {"VALIDATION_INVALID_HEADPHONE", ErrorCode::VALIDATION_INVALID_HEADPHONE},
    {"ERR_UNSUPPORTED_MODE", ErrorCode::ERR_UNSUPPORTED_MODE},

    // Internal
    {"INTERNAL_UNKNOWN", ErrorCode::INTERNAL_UNKNOWN},
};

InnerError::InnerError(ErrorCode code, const std::string& message)
    : cpp_code(errorCodeToHex(code)), cpp_message(message) {}

const char* errorCodeToString(ErrorCode code) {
    auto it = kErrorCodeStrings.find(code);
    if (it != kErrorCodeStrings.end()) {
        return it->second;
    }
    return "UNKNOWN_ERROR";
}

const char* getErrorCategory(ErrorCode code) {
    if (code == ErrorCode::OK) {
        return "ok";
    }
    if (isAudioError(code)) {
        return "audio_processing";
    }
    if (isDacError(code)) {
        return "dac_alsa";
    }
    if (isIpcError(code)) {
        return "ipc_zeromq";
    }
    if (isGpuError(code)) {
        return "gpu_cuda";
    }
    if (isValidationError(code)) {
        return "validation";
    }
    return "internal";
}

int toHttpStatus(ErrorCode code) {
    auto it = kHttpStatusMap.find(code);
    if (it != kHttpStatusMap.end()) {
        return it->second;
    }
    return 500;  // Default to Internal Server Error
}

std::string errorCodeToHex(ErrorCode code) {
    std::ostringstream oss;
    oss << "0x" << std::hex << std::setfill('0') << std::setw(4) << static_cast<uint32_t>(code);
    return oss.str();
}

ErrorCode stringToErrorCode(const std::string& str) {
    auto it = kStringToErrorCode.find(str);
    if (it != kStringToErrorCode.end()) {
        return it->second;
    }
    return ErrorCode::INTERNAL_UNKNOWN;
}

}  // namespace AudioEngine
